### DISCUSSION TITLE
Mixed questions regarding best ways to incorporate non-physical 3D and 2D UI elements into collision detection

### DISCUSSION THREAD

**Discussion by KTRosenberg** [jrouwe/JoltPhysics#376] [Q&A]
I'm considering using Jolt, but I have some questions about integrating non-physical objects. 

My application is an interactive system that has moving physics objects, as well as UI elements and objects that are manually teleported and have their positions set programmatically. UI elements would be 2D buttons that appear and disappear. To my knowledge, a lot of physics systems dislike when you start teleporting bodies like this.

Looking at the examples, I see that there are filters I could use to separate different types of objects and decide at runtime whether they should interact at all. Is there a way I could use Jolt for selectively doing collision detection on my non-physical objects in a more programmatic way that, for example, might ignore the physics step altogether in favor of teleportation or a 0 response?
I think the answer is "yes," but I haven't seen an example like this yet.

Another thing I'd be interested-in is per-object gravity, but I see that there's a global value that is queried and applied.

I also wanted to ask how possible it would be to change and update objects' meshes/locations programmatically? For example, most of my objects have dynamic shapes.

Lastly, since I raise the issue of 2D elements, I did find this [discussion of 2D/2.5D support](https://github.com/jrouwe/JoltPhysics/discussions/180) . It looks like things are still experimental when it comes to 2D support, but it'd be great to have it! I am aware of solutions like Box2D and Chipmunk, but everything I hear suggests the former is not well-optimized for today's hardware, and Chipmunk lacks continuous collision detection.
Being able to use a single physics engine for 2D and 3D would bring a lot of flexibility.

Thank you for your time.

**Comment by jrouwe** [comment]
> My application is an interactive system that has moving physics objects, as well as UI elements and objects that are manually teleported and have their positions set programmatically. UI elements would be 2D buttons that appear and disappear. To my knowledge, a lot of physics systems dislike when you start teleporting bodies like this.

I don't fully understand this. Do your UI elements need to exist in the physics world at all? When I read your questions it seems like the UI elements should not have a physics Body because you don't want to have any physics interaction with them at all.

> Another thing I'd be interested-in is per-object gravity, but I see that there's a global value that is queried and applied.

BodyCreationSettings::mGravityFactor is your friend.

> I also wanted to ask how possible it would be to change and update objects' meshes/locations programmatically? For example, most of my objects have dynamic shapes.

Any body can be moved using BodyInterface::SetPositionAndRotation, but if that makes the body intersect with other dynamic bodies then there's a chance that those bodies get stuck (if a dynamic body ends up in the middle of a MeshShape due to the MeshShape teleporting then usually it's hard to figure out what way is 'outside'). If the body is a dynamic body you would usually use BodyInterface::MoveKinematic to move the body over time and allow it to push other objects aside.

> Lastly, since I raise the issue of 2D elements, I did find this [discussion of 2D/2.5D support](https://github.com/jrouwe/JoltPhysics/discussions/180) . 

I answered to your question in that thread.

**Comment by KTRosenberg** [comment]
From my understanding, if I want collision detection at all, I need to add them to the physics world. There’s no option just to have the collisions. (Please correct me if I’m wrong.) The idea here is that these UI objects are in screen-space, but my game/system is a little bit weird in the sense that any object can change its type or have multiple types — so a button in screen space might switch over to being a physics object. It’s a sandbox sort of simulation in a way.

That sort of brings up the question of whether I can have objects in different coordinate systems. My understanding is yes — by putting them in different (groups?)

Thanks for pointing me to `mGravityFactor`

Ah, so I meant that the shapes themselves can change. As in, I might paint more vertices onto a surface at real time to build a mesh. My understanding is that I might want to per-frame remove or edit the shape somehow though.
As for dynamic/static/kinematic, can I change the type dynamically? As I was looking through, I think I did see it was possible.

Thanks for your time (and for the clarification of the 2D case).

**Reply by jrouwe** [reply]
> From my understanding, if I want collision detection at all, I need to add them to the physics world. There’s no option just to have the collisions. (Please correct me if I’m wrong.) The idea here is that these UI objects are in screen-space, but my game/system is a little bit weird in the sense that any object can change its type or have multiple types — so a button in screen space might switch over to being a physics object. It’s a sandbox sort of simulation in a way.

You can also do collision detection on shapes (e.g. Shape::CastRay), they don't need to be in the PhysicsSystem for that to work. But if you have a large number of UI objects that you want to keep track of then it's probably better to place Bodies in a PhysicsSystem. 

> That sort of brings up the question of whether I can have objects in different coordinate systems. My understanding is yes — by putting them in different (groups?)

Using different coordinate systems in the same PhysicsSystem is confusing, but possible: You can create different ObjectLayers for the different groups and you can configure them to not collide through the ObjectLayerFilter. Make sure you also put these layers in different BroadPhaseLayers too so that the objects using different coordinate systems go to a different broadphase tree.

> Ah, so I meant that the shapes themselves can change. As in, I might paint more vertices onto a surface at real time to build a mesh. My understanding is that I might want to per-frame remove or edit the shape somehow though.

The only way to change the shape of a body is to create a new one and then use BodyInterface::SetShape, see ChangeShapeTest. Shapes are non-mutable (with the exception of the MutableCompoundShape).

> As for dynamic/static/kinematic, can I change the type dynamically? As I was looking through, I think I did see it was possible.

This is also possible, see: ChangeMotionTypeTest

**Reply by KTRosenberg** [reply]
Thanks. This all makes sense. Also, I would have put UI objects into a separate PhysicsSystem (probably as kinematic types), but my understanding was that it would be better to use one PhysicsSystem.

One last note: Is creating and setting a new shape per-frame a reasonable operation if I keep and update old vertices separately? This seems to be the only way to support quickly-changing procedural geometry, based on what you described, unless I use MutableCompoundShape.

**Reply by jrouwe** [reply]
> Thanks. This all makes sense. Also, I would have put UI objects into a separate PhysicsSystem (probably as kinematic types), but my understanding was that it would be better to use one PhysicsSystem.

Multiple PhysicsSystems is also a valid option, it may be slightly more expensive to update multiple physics systems instead of one.

> One last note: Is creating and setting a new shape per-frame a reasonable operation if I keep and update old vertices separately? This seems to be the only way to support quickly-changing procedural geometry, based on what you described, unless I use MutableCompoundShape.

If you want to animate vertices then creating a new mesh shape every frame is the only option right now (MutableCompound contains sub shapes, not vertices). Of course you can also create your own Shape type but that's quite a bit of work.

**Reply by KTRosenberg** [reply]
Right, but is it a generally-cheap operation to create a shape in this way?

**Reply by jrouwe** [reply]
No, creating a mesh shape is not cheap, but if your mesh is not too complex it is probably fast enough. All physics engines I know of assume that mesh shapes don't change at run-time and create a collision detection acceleration structure during shape creation. If you change vertices, this acceleration structure needs to be rebuilt, so there's no shortcuts here (unless as said you create your own shape with your own acceleration structure that trades collision detection speed for mesh creation speed or that has some specific knowledge about the kind of modifications you do). Again, changing mesh shapes during simulation is usually not a great idea as dynamic objects can get stuck, so there's not much reason for supporting mutable mesh shapes. For Jolt's MeshShape, what takes most time is to convert a TriangleList to a VertexList and IndexedTriangleList so if you directly supply the latter then it will be a lot faster.

**Reply by KTRosenberg** [reply]
I see. If creating vertices and indices isn't that complicated, it might be worth it for me to generate those then.
Otherwise, I can imagine finding some ways not to do dynamic mesh generation that often.

In any case, thank you so much for answering all of my questions! I appreciate your time.
